Feature/vmp 122 prediction integration

Kamyki opened the pull request.

Kamyki committed:
Add vehicle presence integration to OTP

Add repurposed HasuraGetter for fetching presence heatmaps.
Add CarPresencePredictor and connect it to AStar
Add basic unit tests
Add runnable for fetching data

Kamyki committed:
Merge branch 'dev-1.x' into feature/VMP-122-prediction-integration

# Conflicts:
#	src/main/java/org/opentripplanner/api/common/RoutingResource.java
#	src/main/java/org/opentripplanner/standalone/CommandLineParameters.java
#	src/main/java/org/opentripplanner/updater/GraphUpdaterConfigurator.java

Kamyki committed:
Decouple api getter from hasura getter

Add integration tests

Kamyki committed:
add limit to itineraries num

AdamWiktor commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -128,6 +128,12 @@
+     * Threshold for deciding when vehicle is assumed to be missing when using vehiclePresencePredictor functionality
+     * Default 0 means it is always present.
+     */
+    public float vehiclePredictionThreshold = 0;
```
`private` with getter and setter or `public` without them

Kamyki replied:
let it be public and wait for the day of refactor

AdamWiktor commented on src/main/java/org/opentripplanner/routing/algorithm/AStar.java:
```diff
@@ -152,29 +153,40 @@ boolean iterate() {
         for (Edge edge : edges) {
 
+
+            if (this.runState.rctx.graph.carPresencePredictor != null &&
```
It all should be inside `RentVehicleEdge#traverse`.

AdamWiktor replied:
This will also make this code testable, so please write tests ;)

MironSz replied:
+2

Kamyki replied:
I recall that once we come to conclusion that we cannot reference graph from inside Edge

Kamyki replied:
oh, we can reference `RoutingContext`. I had wrong assumption at the start.

AdamWiktor replied:
At first I thought that we cannot reference Graph, but then I realized, that there is a reference in RoutingContext

AdamWiktor commented on src/main/java/org/opentripplanner/prediction_client/VehiclePresenceGetter.java:
```diff
@@ -0,0 +1,56 @@
+        return Optional.ofNullable(response).map(this::mapVehiclePresence);
+    }
+
+    VehiclePresence mapVehiclePresence(VehiclePresence vehiclePresence) {
```
This function is copied from `VehiclePresenceMapper`

Kamyki replied:
I will remove Mapper class as it is redundant - artefact from times when I tired to conform to `HasuraGetter` interface

AdamWiktor commented on src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicle_presence/VehiclePresenceGraphWriterRunnable.java:
```diff
@@ -0,0 +1,29 @@
+    public void run(Graph graph) {
+        LOG.info("Updating vehicle presence prediction heatmaps from API");
+        if (graph.carPresencePredictor != null) {
+            graph.carPresencePredictor.updateVehiclePresenceHeatmap(vehiclePresenceHeatmapsFromApi);
```
We should create new presence predictor and just switch references as it should be atomic operation

AdamWiktor commented on src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicle_presence/CarPresencePredictor.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Nullable
+    private STRtree chooseIndex(long time) {
+        Instant pointInTime = Instant.ofEpochSecond(time);
```
This is generally really nice code, but I am afraid it could be potentially not efficient enough? Not sure though, as we have <5k vehicles in whole city. Maybe you could precalculate `this.time + 7*60+30`, `this.time + 22*60+30` and `this.time + 37*60+30` and compare it with argument using `<`? I would call those `indexThreshold15`, ... and store them as class fields

Kamyki replied:
I simplified it, but as for using `<` operator am I against. I bet JIT compiler will spiffily this part of code to simple operation and `.isAfter` is more descriptive.

MironSz replied:
I agree with Kamyki

AdamWiktor commented on src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicle_presence/VehiclePresenceGraphWriterRunnable.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void run(Graph graph) {
+        LOG.info("Updating vehicle presence prediction heatmaps from API");
+        if (graph.carPresencePredictor != null) {
```
i guess this is not really necessary, as you create it using `ifPresent`

Kamyki replied:
+1

AdamWiktor commented on src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicle_presence/VehiclePresencePredictionUpdater.java:
```diff
@@ -0,0 +1,65 @@
+        LOG.info("Polling Vehicle Presence Prediction from API");
+        Optional<VehiclePresence> vehiclePresenceHeatmap = vehiclePresenceGetter.getPrediction(url, params);
+        LOG.info("Got vehicle presence map");
+        vehiclePresenceHeatmap.ifPresent(it -> graphUpdaterManager.execute(new VehiclePresenceGraphWriterRunnable(it)));
```
Maybe some or else with logging?

Kamyki replied:
+1

AdamWiktor commented on src/test/java/org/opentripplanner/updater/vehicle_sharing/vehicle_presence/CarPresencePredictorTest.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Collections;
+import java.util.List;
+
+public class CarPresencePredictorTest extends TestCase {
```
junit4 please

Kamyki replied:
I don't remember what happened here :) I recall trying to make parametrised test

AdamWiktor commented on src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicle_presence/CarPresencePredictor.java:
```diff
@@ -0,0 +1,99 @@
+     * @return probability value in range [0,1] indicating presence of vehicle in given time.
+     */
+    public double predict(VehicleDescription vehicleDescription, long time) {
+        if (VehicleType.CAR.name().equalsIgnoreCase(vehicleDescription.getVehicleType().name())) {
```
why not just compare enums?

Kamyki replied:
lapse of judgement

MironSz commented on src/main/java/org/opentripplanner/prediction_client/VehiclePresence.java:
```diff
@@ -0,0 +1,184 @@
+
+public class VehiclePresence {
+
+    private String vehicle;
```
By reading this class, I couldn't determine what "vehicle" represents. Can you change the name to sth more expressive?

Kamyki replied:
At the cost of annotation for deserialisation. Since it is DTO class

MironSz commented on src/main/java/org/opentripplanner/prediction_client/VehiclePresence.java:
```diff
@@ -0,0 +1,184 @@
+        public Prediction(double lon, double lat, double value, Envelope envelope) {
+            this.lon = lon;
+            this.lat = lat;
+            this.value = value;
```
Can you change "value" to something more expressive?

Kamyki replied:
what exactly? For me value in class Prediction is self explanatory...

Kamyki committed:
Code review changes

Move renting and prediction logic to RentVehicleEdge from AStar
Add test for RentingVehicleEdge in prediction mode
Style changes
Create new prediction each time instead of updating

Kamyki committed:
Add descriptive comments and names in key areas

Kamyki committed:
Extend timeout on prediction api to 2min

Kamyki committed:
Change name of predictionVehicleApi to predictionApiUrl

Kamyki committed:
Merge branch 'dev-1.x' into feature/VMP-122-prediction-integration
